content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md

```diff
@@ -0,0 +1,109 @@
+
+- `--feature-gates=ControllerManagerLeaderMigration=true` to enable Leader Migration as an alpha feature
+- `--enable-leader-migration` to enable Leader Migration on the controller manager
+- `--leader-migration-config=/etc/leadermigration.conf` to set configuration file
+
+Restart `kube-controller-manager` on each node. At this moment, `kube-controller-manager` has leader migration enabled and is ready for the migration.
+
+### Deploy Cloud Controller Manager
+
+The desired state of controller-to-manager assignment can be represented by a new configuration file, shown as follows. Please note `component` field of each `controllerLeaders` changing from `kube-controller-manager` to `cloud-controller-manager`.
+
+```yaml
+kind: LeaderMigrationConfiguration
+apiVersion: controllermanager.config.k8s.io/v1alpha1
+leaderName: cloud-provider-extraction-migration
+resourceLock: leases
+controllerLeaders:
+  - name: route
+    component: cloud-controller-manager
+  - name: service
+    component: cloud-controller-manager
+  - name: cloud-node-lifecycle
+    component: cloud-controller-manager
+
+```
+
+When creating new control plane nodes, the content should be deploy to `/etc/leadermigration.conf`. The manifest of `cloud-controller-manager` should be updated to mount the configuration file in the same manner as `kube-controller-manager`. Similarly, add `--feature-gates=ControllerManagerLeaderMigration=true`,`--enable-leader-migration`, and `--leader-migration-config=/etc/leadermigration.conf` to the arguments.
+
+Create a new control plane node with the updated `cloud-controller-manager` manifest, and with the `--cloud-provider` flag unset for `kube-controller-manager`. The newly deployed `kube-controller-manager` MUST NOT have Leader Migration enabled because it does not run the migrated controllers with an external cloud provider and thus is not involved in the migration.
+
+Please refer to [Cloud Controller Manager Administration](/docs/tasks/administer-cluster/running-cloud-controller/) for more detail on how to deploy `cloud-controller-manager`.
+
+### Upgrade Control Plane
+
+The control plane now contains nodes of both versions. The old version running `kube-controller-manager` only, and the new one running both `kube-controller-manager` and `cloud-controller-manager`. The migrated controllers, as specified in the configuration, are running under either old `kube-controller-manager` or `cloud-controller-manager` but not both at any time, depending on which controller manager holds the migration lease.
+
+In a rolling manner, create a new control plane node and bring down one of the old until the control plane contains only nodes of the new version.
+
+### (Optional) Disable Leader Migration
+
+Now that the control plane has been upgraded to run both `kube-controller-manager` and `cloud-controller-manager`, Leader Migration has finished its job and can be safely disabled to save one lease resource.
```

Comment:
It might be needed again if a rollback needs to happen. The rollback might be required for some completely unrelated problem. Maybe recommend that cluster administrators wait an additional release before doing this?

Reply:
I think "an additional release" part heavily depends on the out-of-tree cloud provider. Probably better let the user decide.

Reply:
Still, I think it would be great to note that a rollback may happen, and Leader Migration can be enabled back any time as the user likes.